DB/src/main/java/io/deephaven/stream/StreamToTableAdapter.java

```diff
@@ -46,11 +143,65 @@ public void close() {
     public void refresh() {
         streamPublisher.flush();
         // Switch columns, update index, deliver notification
+
+        final long oldSize = index.size();
+        final long newSize;
+
+        synchronized (this) {
+            newSize = bufferChunkSources == null ? 0 : bufferChunkSources[0].getSize();
+
+            if (oldSize == 0 && newSize == 0) {
+                return;
+            }
+
+            ChunkColumnSource<?> [] oldPrev = prevChunkSources;
+
+            for (int ii = 0; ii < switchSources.length; ++ii) {
+                if (prevChunkSources != null) {
+                    prevChunkSources[ii].clear();
+                }
+            }
+
+            if (bufferChunkSources == null) {
+                // null out our current values
+                for (int ii = 0; ii < switchSources.length; ++ii) {
+                    //noinspection unchecked
+                    switchSources[ii].setNewCurrent((ColumnSource) nullColumnSources[ii]);
+                }
+            } else {
+                for (int ii = 0; ii < switchSources.length; ++ii) {
+                    switchSources[ii].setNewCurrent((ColumnSource) bufferChunkSources[ii]);
+                }
+            }
+            // current becomes prev, buffer becomes current, buffer is set to cleared prev for the next pass
+            prevChunkSources = currentChunkSources;
+            currentChunkSources = bufferChunkSources;
+            bufferChunkSources = oldPrev;
+
+            if (oldSize < newSize) {
+                index.insertRange(oldSize, newSize - 1);
+            } else if (oldSize > newSize) {
+                index.removeRange(newSize, oldSize - 1);
+            }
+        }
+
+        table.notifyListeners(new ShiftAwareListener.Update(Index.FACTORY.getFlatIndex(newSize), Index.FACTORY.getFlatIndex(oldSize), Index.FACTORY.getEmptyIndex(), IndexShiftData.EMPTY, ModifiedColumnSet.EMPTY));
     }
 
     @SafeVarargs
     @Override
     public final void accept(@NotNull final WritableChunk<Attributes.Values>... data) {
         // Accumulate data into buffered column sources
+        synchronized (this) {
+            if (bufferChunkSources == null) {
+                bufferChunkSources = makeChunkSources(columnTypesArray, columnNamesArray);
+            }
+            if (data.length != bufferChunkSources.length) {
+                throw new IllegalStateException();
```

Comment:
What state is illegal? (exception text)

Also consider that IllegalState percolating up makes the worker bail/exit; UncheckedDeephavenException is what you can use when you want the exception to get to the toplevel, notify the user, abort whatever the current thing is, but not kill the worker etc.

Maybe this code is running in the KafkaIngester thread tho?   In that case maybe a TODO with the new issue linked for the error handling stuff we discussed.

In case is useful, for your convenience, that issue's link: https://github.com/deephaven/deephaven-core/issues/934